client/src/explorer_common/abstract_explorer_scheme.js

```diff
@@ -0,0 +1,439 @@
+      });
+
+    return {
+      set_query,
+      toggle_node,
+      clear_query,
+      enable_loading,
+      expand_all,
+      collapse_all,
+      clear_expanded_collapsed,
+    };
+  }
+
+  get_base_hierarchy_selector() {
+    throw "NotImplemented";
+  }
+
+  @cached_property
+  get_is_filtering_selector() {
+    return (state) => state.root.query.length > 3;
+  }
+
+  @cached_property
+  get_query_filter_func_selector() {
+    // note that we create a new filter FUNCTION each time the base hierarchy changes
+    // this filter function's closure keeps a processed version of the hierarchy
+    // this keeps successive searches acrross different queries quick
+
+    return createSelector([this.get_base_hierarchy_selector()], (base_h7y) =>
+      substr_search_generator(base_h7y)
+    );
+  }
+
+  @cached_property
+  get_query_filtered_hierarchy_selector() {
+    return createSelector(
+      [
+        this.get_base_hierarchy_selector(),
+        this.get_query_filter_func_selector(),
+        (state) => state.root.query,
+      ],
+      (base_h7y, query_filter_func, query) => {
+        if (query.length < 4) {
+          return base_h7y;
+        } else {
+          const filtered = filter_hierarchy(
+            base_h7y,
+            query_filter_func(query),
+            { markSearchResults: true }
+          );
+          ensureVisibility(filtered, _.property("is_search_match"));
+          return filtered;
+        }
+      }
+    );
+  }
+
+  @cached_property
+  get_filter_func_selector() {
+    // override to filter based on scheme state
+    return (_state) => {
+      return (node) => node;
+    };
+  }
+
+  @cached_property
+  get_sort_func_selector() {
+    // override to filter based on scheme state
+    // unlike functions passed to _.sortBy, this function takes a list and must return a new sorted list
+    return (state) => {
+      return (nodes) => nodes;
+    };
+  }
+
+  @cached_property
+  get_fully_filtered_hierarchy_selector() {
+    const get_query_filtered_hierarchy = this.get_query_filtered_hierarchy_selector();
+    const get_filter_func = this.get_filter_func_selector();
+    return createSelector(
+      [get_query_filtered_hierarchy, get_filter_func],
+      (query_filtered_hierarchy, filter_func) => {
+        return filter_func(query_filtered_hierarchy);
+      }
+    );
+  }
+
+  @cached_property
+  get_sorted_filtered_hierarchy_selector() {
+    const get_fully_filtered_hierarchy = this.get_fully_filtered_hierarchy_selector();
+    const get_sort_func = this.get_sort_func_selector();
+    return createSelector(
+      [get_fully_filtered_hierarchy, get_sort_func],
+      (filtered_hierarchy, sort_func) => {
+        return _.chain(filtered_hierarchy)
+          .pipe((h7y) => sort_hierarchy(h7y, sort_func))
+          .sortBy(negative_search_relevance_func) //search results always take precedence
+          .value();
+      }
+    );
+  }
+
+  _get_flat_nodes(state) {
+    let flat_nodes;
```

Comment:
Modules with top-level (as in, on-import) side effects need to be named `<name>.side-effects.js` or webpack will most likely break them during optimization.

You might be thinking of the global declarations, I gathered those all up in one place because I wanted to discourage their use... they're bad for tree shaking, but also just bad in general!

Also it's a holiday, let's all log off, haha.

Reply:
Any side-effects here are only persisted on _instances_ of this class, and those instances will get created and garbage-collected within individual react component lifecycles, so there's no side-effects at the module-level.